py_ecc/fields/field_elements.py

```diff
@@ -0,0 +1,366 @@
+from typing import (
+    cast,
+    List,
+    Sequence,
+    Union,
+)
+
+
+from py_ecc.fields.field_properties import (
+    field_properties,
+)
+
+from py_ecc.utils import (
+    deg,
+    poly_rounded_div,
+    prime_field_inv,
+)
+
+
+IntOrFQ = Union[int, "FQ"]
+
+
+class FQ(object):
+    """
+    A class for field elements in FQ. Wrap a number in this class,
+    and it becomes a field element.
+    """
+    n = None  # type: int
+    field_modulus = None
+    # curve_name can be either 'bn128' or 'bls12_381'
+    # This is needed to obtain field_modulus, FQ2_MODULUS_COEFFS
+    # and FQ12_MODULUS_COEFFS from the curve properties
+    curve_name = None
+
+    def __init__(self, val: IntOrFQ) -> None:
+        if self.curve_name is None:
+            raise AttributeError("Curve Name hasn't been specified")
+
+        self.field_modulus = field_properties[self.curve_name]["field_modulus"]
+
+        if isinstance(val, FQ):
+            self.n = val.n
+        elif isinstance(val, int):
+            self.n = val % self.field_modulus
+        else:
+            raise TypeError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(val))
+            )
+
+    def __add__(self, other: IntOrFQ) -> "FQ":
+        if isinstance(other, FQ):
+            on = other.n
+        elif isinstance(other, int):
+            on = other
+        else:
+            raise TypeError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(other))
+            )
+
+        return type(self)((self.n + on) % self.field_modulus)
+
+    def __mul__(self, other: IntOrFQ) -> "FQ":
+        if isinstance(other, FQ):
+            on = other.n
+        elif isinstance(other, int):
+            on = other
+        else:
+            raise TypeError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(other))
+            )
+
+        return type(self)((self.n * on) % self.field_modulus)
+
+    def __rmul__(self, other: IntOrFQ) -> "FQ":
+        return self * other
+
+    def __radd__(self, other: IntOrFQ) -> "FQ":
+        return self + other
+
+    def __rsub__(self, other: IntOrFQ) -> "FQ":
+        if isinstance(other, FQ):
+            on = other.n
+        elif isinstance(other, int):
+            on = other
+        else:
+            raise TypeError(
```

Comment:
This sequence of `if/elif/else` is repeated multiple times.  What do you think about.

```python
def normalize_point(value: IntOrFQ) -> int:
    if isinstance(value, int):
        return int
    elif isinstance(value, FQ):
        return int
    else:
        raise ...
```

Reply:
@pipermerriam there are many places where `if/elif/else` pattern is being used. But out of them, I feel that very few come under the pattern that you are suggesting (i.e. where the usage of `normalize_point` is apt).

But if you feel that this would make the code more readable, please let me know. Will make the changes.

Reply:
- https://github.com/ethereum/py_ecc/pull/41/files/9c80ebd6ea0b461d978565b4492cd2c4f40ba5a6#diff-14f5ca34559da438b521e765e0cb1fdaR52
- https://github.com/ethereum/py_ecc/pull/41/files/9c80ebd6ea0b461d978565b4492cd2c4f40ba5a6#diff-14f5ca34559da438b521e765e0cb1fdaR65
- https://github.com/ethereum/py_ecc/pull/41/files/9c80ebd6ea0b461d978565b4492cd2c4f40ba5a6#diff-14f5ca34559da438b521e765e0cb1fdaR84

Maybe my intent wasn't clear.  The `if/elif/else` block in all of these three locations could be replaced by:

```
def __rsub(self, other: IntOrFQ) -> "FQ":
    value = normalize_point(other)
    return type(self)((value - self.n) % self.field_modulus)
```

It makes for much more concise/easy-to-read methods IMHO.